In [1]:
import numpy as np
from sklearn.datasets import fetch_openml
from sklearn.model_selection import cross_val_score, GridSearchCV
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.svm import SVC
import pandas as pd

1. Support vector machines using rbf-kernels perform very well on the MNIST dataset. By tuning your
parameters you should be able to get over 95% test accuracy. So, the first part of this exercise is to find C
and gamma to obtain that kind of scores. You may use a smaller part of MNIST for training and still obtain
good scores. Recall that the hyperparameters have to be found without laying your hands on the test set,
i.e. use either cross-validation, a validation set or some other technique to distinguish between different
models. Report in your code as comments, or in a separate document, the grid (or whatever technique for
hyperparameter search your are using) which was searched and the resulting best hyperparameters.


In [2]:
mnist = fetch_openml('mnist_784', version=1)

In [3]:
X, y = mnist.data.values, mnist.target.values
random = np.random.randint(1, 1001)

In [4]:
X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=60000, random_state=random)
X_train_sample, _, y_train_sample, _ = train_test_split(X_train, y_train, train_size=2000, random_state=random)
# take a sample of 10000 points from X_train and y_train

In [5]:
# create a clf rbf kernel
scaler = StandardScaler()
X_train_sample_scaled = scaler.fit_transform(X_train_sample)
clf = SVC()
param_grid = [{
    "kernel": ["rbf"],
    "C": [4.89 , 4.92, 5, 5.5, 6],
    "gamma": [0.001, 0.0008, 0.0009]
}]


grid_search = GridSearchCV(clf, param_grid, cv=3, n_jobs=-1, scoring='accuracy', return_train_score=True)
grid_search.fit(X_train_sample_scaled, y_train_sample)

GridSearchCV(cv=3, estimator=SVC(), n_jobs=-1,
             param_grid=[{'C': [4.89, 4.92, 5, 5.5, 6],
                          'gamma': [0.001, 0.0008, 0.0009],
                          'kernel': ['rbf']}],
             return_train_score=True, scoring='accuracy')

After running the grid search multiple times, I have narrowed down the best parameters for my model. All parameters give similar results, so I believe I will have to manually fine tune them to get the best results.

In [6]:
# printing all results
cvres = grid_search.cv_results_
df = pd.DataFrame(cvres)
df

mean_fit_time  std_fit_time  mean_score_time  std_score_time param_C  \
0        0.571102      0.097506         1.442512        0.165270    4.89   
1        0.544100      0.017175         1.435506        0.059097    4.89   
2        0.981664      0.030134         2.477553        0.101715    4.89   
3        0.847322      0.125826         1.852221        0.411925    4.92   
4        0.993500      0.582772         1.710542        0.478700    4.92   
5        0.620441      0.214917         1.589848        0.604680    4.92   
6        0.530020      0.043287         1.500848        0.102906       5   
7        0.697005      0.160517         1.532127        0.027471       5   
8        1.313558      0.611022         1.395386        0.117386       5   
9        0.565885      0.071132         1.657959        0.043765     5.5   
10       1.092020      0.555166         1.494159        0.063276     5.5   
11       1.191121      0.312989         1.276927        0.075452     5.5   
12       1.550957      0.289478         1.312229        0.154264       6   
13       0.906454      0.249002         1.127714        0.109248       6   
14       0.642234      0.125414         1.190262        0.025443       6   

   param_gamma param_kernel                                         params  \
0        0.001          rbf   {'C': 4.89, 'gamma': 0.001, 'kernel': 'rbf'}   
1       0.0008          rbf  {'C': 4.89, 'gamma': 0.0008, 'kernel': 'rbf'}   
2       0.0009          rbf  {'C': 4.89, 'gamma': 0.0009, 'kernel': 'rbf'}   
3        0.001          rbf   {'C': 4.92, 'gamma': 0.001, 'kernel': 'rbf'}   
4       0.0008          rbf  {'C': 4.92, 'gamma': 0.0008, 'kernel': 'rbf'}   
5       0.0009          rbf  {'C': 4.92, 'gamma': 0.0009, 'kernel': 'rbf'}   
6        0.001          rbf      {'C': 5, 'gamma': 0.001, 'kernel': 'rbf'}   
7       0.0008          rbf     {'C': 5, 'gamma': 0.0008, 'kernel': 'rbf'}   
8       0.0009          rbf     {'C': 5, 'gamma': 0.0009, 'kernel': 'rbf'}   
9        0.001          rbf    {'C': 5.5, 'gamma': 0.001, 'kernel': 'rbf'}   
10      0.0008          rbf   {'C': 5.5, 'gamma': 0.0008, 'kernel': 'rbf'}   
11      0.0009          rbf   {'C': 5.5, 'gamma': 0.0009, 'kernel': 'rbf'}   
12       0.001          rbf      {'C': 6, 'gamma': 0.001, 'kernel': 'rbf'}   
13      0.0008          rbf     {'C': 6, 'gamma': 0.0008, 'kernel': 'rbf'}   
14      0.0009          rbf     {'C': 6, 'gamma': 0.0009, 'kernel': 'rbf'}   

    split0_test_score  split1_test_score  split2_test_score  mean_test_score  \
0            0.899550           0.899550           0.927928         0.909009   
1            0.907046           0.899550           0.929429         0.912009   
2            0.902549           0.895052           0.929429         0.909010   
3            0.899550           0.899550           0.927928         0.909009   
4            0.907046           0.899550           0.929429         0.912009   
5            0.902549           0.895052           0.929429         0.909010   
6            0.899550           0.899550           0.926426         0.908509   
7            0.907046           0.899550           0.929429         0.912009   
8            0.902549           0.895052           0.929429         0.909010   
9            0.901049           0.901049           0.926426         0.909508   
10           0.910045           0.899550           0.927928         0.912508   
11           0.902549           0.898051           0.926426         0.909009   
12           0.901049           0.902549           0.927928         0.910509   
13           0.910045           0.901049           0.927928         0.913007   
14           0.902549           0.899550           0.926426         0.909508   

    std_test_score  rank_test_score  split0_train_score  split1_train_score  \
0         0.013377               12            0.998500            0.996999   
1         0.012693                3            0.996249            0.995499   
2         0.014759  

In [16]:
random = np.random.randint(1, 1001)
X_train_sample, _, y_train_sample, _ = train_test_split(X_train, y_train, train_size=10000, random_state=random)
X_train_sample_scaled = scaler.fit_transform(X_train_sample)
clf_best = SVC(C=4.92, gamma=0.001)
clf_best.fit(X_train_sample_scaled, y_train_sample)
test_score = clf_best.score(scaler.transform(X_test), y_test)
print("Test Accuracy: {:.2f}%".format(test_score * 100))

Test Accuracy: 95.18%


- After some fine tuning of the C regularization parameter, I was able to achieve a test accuracy of 95.18% on the test set.
- The best parameters for my model seems to be C = 4.89 and gamma = 0.001

2. The second part of this exercise is to compare the built-in binarization scheme used for the SVC class,
namely one-vs-one, against the one-vs-all scheme, which was discussed in Lecture 5. You should implement
your own version of one-vs-all SVM and compare your results against the built in version. To make the
comparison simple you should keep the same hyperparameters which you found in the first part of this
exercise. Which was the best classifier? If studying the confusion matrix was there any apparent difference
between the two methods in terms of misclassifications?

In [1]:
import Functions as f

zz = f.one_vs_all(X_train_sample_scaled, y_train_sample, X_test, C=4.89, gamma=0.001)

NameError: name 'X_train_sample_scaled' is not defined